src/client/Microsoft.Identity.Client/Platforms/Features/WamBroker/AadPlugin.cs

```diff
@@ -27,25 +30,80 @@ public AadPlugin(IWamProxy wamProxy, IWebAccountProviderFactory webAccountProvid
             _logger = logger;
         }
 
-        public async Task<IEnumerable<IAccount>> GetAccountsAsync(string clientID)
+        /// <summary>
+        /// The algorithm here is much more complex in order to workaround a limitation in the AAD plugin's 
+        /// handling of guest accounts: 
+        /// 
+        /// 1. Read the accounts from WAM.AADPlugin
+        /// 2. For each account, we need to find its home_account_id as the one from WAM may not be correct
+        /// 3. If we can find a cached account with the same LocalAccountId or UPN, use it
+        /// 4. If not, make a simple silent token request and use the client info provided
+        /// </summary>
+        public async Task<IReadOnlyList<IAccount>> GetAccountsAsync(
+            string clientId,
+            string authority,
+            Cache.ICacheSessionManager cacheSessionManager,
+            Instance.Discovery.IInstanceDiscoveryManager instanceDiscoveryManager)
         {
             var webAccountProvider = await _webAccountProviderFactory.GetAccountProviderAsync("organizations").ConfigureAwait(false);
+            var wamAccounts = await _wamProxy.FindAllWebAccountsAsync(webAccountProvider, clientId).ConfigureAwait(false);
 
-            var webAccounts = await _wamProxy.FindAllWebAccountsAsync(webAccountProvider, clientID).ConfigureAwait(false);
-
-            var msalAccounts = webAccounts
-                .Select(webAcc => ConvertToMsalAccountOrNull(webAcc))
-                .Where(a => a != null)
-                .ToList();
+            if (wamAccounts.Count > 0)
+            {
+                var webAccountEnvs = wamAccounts
+                    .Select(w =>
+                    {
+                        _wamProxy.TryGetAccountProperty(w, "Authority", out string accountAuthority);
+                        if (accountAuthority != null)
+                        {
+                            return (new Uri(accountAuthority)).Host;
+                        }
+                        else
+                        {
+                            _logger.WarningPii(
+                                $"[WAM AAD Provider] Could not convert the WAM account {w.UserName} (id: {w.Id}) to an MSAL account because the Authority could not be found",
+                                $"[WAM AAD Provider] Could not convert the WAM account {w.Id} to an MSAL account because the Authority could not be found");
+
+                            return null;
+                        }
+                    })
+                    .Where(a => a != null);
+
+                var instanceMetadata = await instanceDiscoveryManager.GetMetadataEntryTryAvoidNetworkAsync(authority, webAccountEnvs, cacheSessionManager.RequestContext)
+                    .ConfigureAwait(false);
```

Comment:
I think you can do this:
`cacheSessionManager.RequestContext.ServiceBundle.InstanceDiscoveryManager.GetMetadataEntryTryAvoidNetworkAsync(authority, webAccountEnvs, cacheSessionManager.RequestContext);`

so you don't have to pass the instanceDiscoveryManager around, you can get it from the RC.

Reply:
I agree, but I like this a bit more because of separation of concerns. RequestContext has a LOT of things inside, so whenever a method requires the RequestContext it's actually very difficult to understand what that method actually needs. This way, I can immediately tell that to GetAccounts, the broker needs to manipulate the cache and to get cloud aliases.